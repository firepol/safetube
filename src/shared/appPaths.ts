// This file contains only interfaces and types that can be shared between main and renderer processes
// The actual AppPaths implementation is in src/main/appPaths.ts

export interface AppPathsInterface {
  getConfigDir(): string;
  getCacheDir(): string;
  getLogsDir(): string;
  getUserDataDir(): string;
  isDev(): boolean;
  getConfigPath(filename: string): string;
  getCachePath(filename: string): string;
  getLogPath(filename: string): string;
}