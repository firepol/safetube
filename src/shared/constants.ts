/**
 * The default admin password hash.
 * The password is 'paren234', hashed with bcrypt.
 */
export const DEFAULT_ADMIN_PASSWORD_HASH = '$2b$10$CD78JZagbb56sj/6SIJfyetZN5hYjICzbPovBm5/1mol2K53bWIWy';
