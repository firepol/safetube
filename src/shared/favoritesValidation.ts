import { z } from 'zod';
import { FavoriteVideo, FavoritesConfig, FavoriteValidationResult } from './types';

// Zod schema for FavoriteVideo validation
export const FavoriteVideoSchema = z.object({
  videoId: z.string().min(1, 'Video ID is required'),
  dateAdded: z.string().datetime({ message: 'Invalid ISO date string for dateAdded' }),
  sourceType: z.enum(['youtube', 'local', 'dlna'], {
    errorMap: () => ({ message: 'Source type must be youtube, local, or dlna' })
  }),
  sourceId: z.string().min(1, 'Source ID is required'),
  title: z.string().min(1, 'Title is required'),
  thumbnail: z.string().optional(),
  duration: z.number().min(0, 'Duration must be non-negative').optional()
});

// Zod schema for FavoritesConfig validation
export const FavoritesConfigSchema = z.object({
  favorites: z.array(FavoriteVideoSchema),
  lastModified: z.string().datetime({ message: 'Invalid ISO date string for lastModified' })
});

// Default favorites configuration
export const DEFAULT_FAVORITES_CONFIG: FavoritesConfig = {
  favorites: [],
  lastModified: new Date().toISOString()
};

/**
 * Validates and sanitizes a FavoriteVideo object
 */
export function validateFavoriteVideo(data: unknown): FavoriteValidationResult {
  try {
    const result = FavoriteVideoSchema.safeParse(data);

    if (!result.success) {
      return {
        isValid: false,
        errors: result.error.issues.map(issue =>
          `${issue.path.join('.')}: ${issue.message}`
        )
      };
    }

    return {
      isValid: true,
      errors: [],
      sanitized: result.data
    };
  } catch (error) {
    return {
      isValid: false,
      errors: [`Validation error: ${error instanceof Error ? error.message : 'Unknown error'}`]
    };
  }
}

/**
 * Validates and sanitizes a FavoritesConfig object
 */
export function validateFavoritesConfig(data: unknown): {
  isValid: boolean;
  errors: string[];
  sanitized?: FavoritesConfig;
} {
  try {
    const result = FavoritesConfigSchema.safeParse(data);

    if (!result.success) {
      return {
        isValid: false,
        errors: result.error.issues.map(issue =>
          `${issue.path.join('.')}: ${issue.message}`
        )
      };
    }

    return {
      isValid: true,
      errors: [],
      sanitized: result.data
    };
  } catch (error) {
    return {
      isValid: false,
      errors: [`Validation error: ${error instanceof Error ? error.message : 'Unknown error'}`]
    };
  }
}

/**
 * Sanitizes video data for adding to favorites
 */
export function sanitizeVideoData(data: Partial<FavoriteVideo>): Partial<FavoriteVideo> {
  return {
    videoId: typeof data.videoId === 'string' ? data.videoId.trim() : undefined,
    dateAdded: data.dateAdded, // ISO string validation handled by schema
    sourceType: data.sourceType, // Enum validation handled by schema
    sourceId: typeof data.sourceId === 'string' ? data.sourceId.trim() : undefined,
    title: typeof data.title === 'string' ? data.title.trim() : undefined,
    thumbnail: typeof data.thumbnail === 'string' ? data.thumbnail.trim() : undefined,
    duration: typeof data.duration === 'number' && data.duration >= 0 ? data.duration : undefined,
  };
}

/**
 * Creates a new FavoriteVideo from video metadata
 */
export function createFavoriteFromVideo(
  videoId: string,
  sourceType: 'youtube' | 'local' | 'dlna',
  sourceId: string,
  title: string,
  thumbnail?: string,
  duration?: number
): FavoriteVideo {
  const favorite: FavoriteVideo = {
    videoId,
    dateAdded: new Date().toISOString(),
    sourceType,
    sourceId,
    title: title.trim(),
    thumbnail: thumbnail?.trim(),
    duration: duration !== undefined ? Math.max(0, duration) : undefined
  };

  const validation = validateFavoriteVideo(favorite);
  if (!validation.isValid) {
    throw new Error(`Invalid favorite video data: ${validation.errors.join(', ')}`);
  }

  return validation.sanitized!;
}