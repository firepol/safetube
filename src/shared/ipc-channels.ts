/**
 * IPC Channel Constants - Single Source of Truth
 *
 * This file defines all IPC channel names used for communication between
 * main process, preload, and renderer processes.
 *
 * Benefits:
 * - Single source of truth for all IPC channels
 * - TypeScript autocomplete and type safety
 * - Refactoring safety - change channel name in one place
 * - Compile-time error detection for typos
 *
 * Usage:
 * - Main process: ipcMain.handle(IPC.FAVORITES.GET_ALL, ...)
 * - Preload: ipcRenderer.invoke(IPC.FAVORITES.GET_ALL)
 */

export const IPC = {
  // ============================================================================
  // DATABASE OPERATIONS
  // ============================================================================

  DATABASE: {
    HEALTH_CHECK: 'database:health-check',
  },

  // ============================================================================
  // FAVORITES (Database-backed)
  // ============================================================================

  FAVORITES: {
    GET_ALL: 'database:favorites:get-all',
    ADD: 'database:favorites:add',
    REMOVE: 'database:favorites:remove',
    IS_FAVORITE: 'database:favorites:is-favorite',
    TOGGLE: 'database:favorites:toggle',
  },

  // ============================================================================
  // VIEW RECORDS (Watch history & resume positions)
  // ============================================================================

  VIEW_RECORDS: {
    GET: 'database:view-records:get',
    UPDATE: 'database:view-records:update',
    GET_HISTORY: 'database:view-records:get-history',
    GET_RECENTLY_WATCHED: 'database:view-records:get-recently-watched',
  },

  // ============================================================================
  // VIDEOS (Database operations)
  // ============================================================================

  VIDEOS: {
    GET_BY_ID: 'database:videos:get-by-id',
    GET_BY_SOURCE: 'database:videos:get-by-source',
    SEARCH: 'database:videos:search',
    UPDATE_METADATA: 'database:videos:update-metadata',
    UPDATE_AVAILABILITY: 'database:videos:update-availability',
    BATCH_UPSERT: 'database:videos:batch-upsert',
  },

  // ============================================================================
  // SOURCES (Video source management)
  // ============================================================================

  SOURCES: {
    GET_ALL: 'database:sources:get-all',
    GET_BY_ID: 'database:sources:get-by-id',
    CREATE: 'database:sources:create',
    UPDATE: 'database:sources:update',
    DELETE: 'database:sources:delete',
  },

  // ============================================================================
  // VIDEO SOURCES (Legacy - File-backed)
  // ============================================================================

  VIDEO_SOURCES: {
    GET_ALL: 'video-sources:get-all',
    SAVE_ALL: 'video-sources:save-all',
    VALIDATE_YOUTUBE_URL: 'video-sources:validate-youtube-url',
    VALIDATE_LOCAL_PATH: 'video-sources:validate-local-path',
  },

  // ============================================================================
  // YOUTUBE CACHE (Database-backed)
  // ============================================================================

  YOUTUBE_CACHE_DB: {
    GET_CACHED_RESULTS: 'database:youtube-cache:get-cached-results',
    SET_CACHED_RESULTS: 'database:youtube-cache:set-cached-results',
    CLEAR_CACHE: 'database:youtube-cache:clear-cache',
    GET_PAGE: 'youtube-cache:get-page',
    SAVE_PAGE: 'youtube-cache:save-page',
    CLEAR_SOURCE: 'youtube-cache:clear-source',
  },

  // ============================================================================
  // YOUTUBE CACHE (Legacy - File-backed)
  // ============================================================================

  YOUTUBE_CACHE: {
    GET: 'youtube-cache:get',
    SET: 'youtube-cache:set',
    CLEAR_EXPIRED: 'youtube-cache:clear-expired',
    LOAD_CONFIG: 'youtube-cache:load-config',
  },

  // ============================================================================
  // TIME TRACKING
  // ============================================================================

  TIME_TRACKING: {
    RECORD_VIDEO_WATCHING: 'time-tracking:record-video-watching',
    GET_TIME_TRACKING_STATE: 'time-tracking:get-time-tracking-state',
    GET_TIME_LIMITS: 'time-tracking:get-time-limits',
  },

  // ============================================================================
  // ADMIN / PARENT ACCESS
  // ============================================================================

  ADMIN: {
    AUTHENTICATE: 'admin:authenticate',
    CHANGE_PASSWORD: 'admin:change-password',
    HASH_PASSWORD: 'admin:hash-password',
    ADD_EXTRA_TIME: 'admin:add-extra-time',
    GET_TIME_EXTRA: 'admin:get-time-extra',
    WRITE_TIME_LIMITS: 'admin:write-time-limits',
    GET_LAST_WATCHED_VIDEO_WITH_SOURCE: 'admin:get-last-watched-video-with-source',
  },

  // ============================================================================
  // VIDEO LOADING & DATA
  // ============================================================================

  VIDEO_LOADING: {
    LOAD_ALL_VIDEOS_FROM_SOURCES: 'load-all-videos-from-sources',
    LOAD_VIDEOS_FROM_SOURCES: 'load-videos-from-sources',
    LOAD_SOURCES_FOR_KID_SCREEN: 'load-sources-for-kid-screen',
    LOAD_VIDEOS_FOR_SOURCE: 'load-videos-for-source',
    GET_VIDEO_DATA: 'get-video-data',
    GET_PAGINATED_VIDEOS: 'get-paginated-videos',
    GET_WATCHED_VIDEOS: 'get-watched-videos',
  },

  // ============================================================================
  // LOCAL FILE OPERATIONS
  // ============================================================================

  LOCAL_FILES: {
    GET_LOCAL_FILE: 'get-local-file',
    GET_LOCAL_FOLDER_CONTENTS: 'get-local-folder-contents',
    GET_LOCAL_SOURCE_VIDEO_COUNT: 'get-local-source-video-count',
    GET_FOLDER_VIDEO_COUNT: 'get-folder-video-count',
    GET_LOCAL_VIDEO_DURATION: 'get-local-video-duration',
  },

  // ============================================================================
  // DLNA OPERATIONS
  // ============================================================================

  DLNA: {
    GET_DLNA_FILE: 'get-dlna-file',
  },

  // ============================================================================
  // VIDEO PLAYBACK
  // ============================================================================

  PLAYBACK: {
    GET_VIDEO_STREAMS: 'get-video-streams',
    GET_PLAYER_CONFIG: 'get-player-config',
  },

  // ============================================================================
  // VIDEO CONVERSION & PROCESSING
  // ============================================================================

  CONVERSION: {
    GET_VIDEO_CODEC_INFO: 'get-video-codec-info',
    NEEDS_VIDEO_CONVERSION: 'needs-video-conversion',
    HAS_CONVERTED_VIDEO: 'has-converted-video',
    GET_EXISTING_CONVERTED_VIDEO_PATH: 'get-existing-converted-video-path',
    GET_CONVERSION_STATUS: 'get-conversion-status',
    START_VIDEO_CONVERSION: 'start-video-conversion',
  },

  // ============================================================================
  // DOWNLOADS
  // ============================================================================

  DOWNLOADS: {
    START: 'download:start',
    GET_STATUS: 'download:get-status',
    CANCEL: 'download:cancel',
    IS_DOWNLOADING: 'download:is-downloading',
    RESET_STATUS: 'download:reset-status',
    CHECK_DOWNLOADED: 'download:check-downloaded',
    CLEANUP: 'downloads:cleanup',
  },

  // ============================================================================
  // DOWNLOADED VIDEOS
  // ============================================================================

  DOWNLOADED_VIDEOS: {
    GET_ALL: 'downloaded-videos:get-all',
    GET_BY_SOURCE: 'downloaded-videos:get-by-source',
    GET_BY_ID: 'downloaded-videos:get-by-id',
    GET_TOTAL_SIZE: 'downloaded-videos:get-total-size',
  },

  // ============================================================================
  // YOUTUBE API
  // ============================================================================

  YOUTUBE: {
    GET_API_KEY: 'get-youtube-api-key',
    GET_VIDEO_INFO: 'get-youtube-video-info',
  },

  // ============================================================================
  // USAGE LOGS (Database-backed)
  // ============================================================================

  USAGE_LOGS: {
    GET_BY_DATE: 'database:usage-logs:get-by-date',
    UPSERT: 'database:usage-logs:upsert',
    INCREMENT: 'database:usage-logs:increment',
    GET_BY_DATE_RANGE: 'database:usage-logs:get-by-date-range',
    GET_MONTHLY: 'database:usage-logs:get-monthly',
  },

  // ============================================================================
  // TIME LIMITS (Database-backed)
  // ============================================================================

  TIME_LIMITS: {
    GET: 'database:time-limits:get',
    UPDATE: 'database:time-limits:update',
    GET_FOR_DAY: 'database:time-limits:get-for-day',
    UPDATE_DAY: 'database:time-limits:update-day',
  },

  // ============================================================================
  // USAGE EXTRAS (Database-backed)
  // ============================================================================

  USAGE_EXTRAS: {
    GET_BY_DATE: 'database:usage-extras:get-by-date',
    GET_TOTAL_MINUTES: 'database:usage-extras:get-total-minutes',
    ADD: 'database:usage-extras:add',
    DELETE: 'database:usage-extras:delete',
  },

  // ============================================================================
  // SETTINGS (Database-backed)
  // ============================================================================

  DB_SETTINGS: {
    GET_SETTING: 'database:settings:get-setting',
    SET_SETTING: 'database:settings:set-setting',
    GET_BY_NAMESPACE: 'database:settings:get-by-namespace',
    SET_BY_NAMESPACE: 'database:settings:set-by-namespace',
  },

  // ============================================================================
  // SETTINGS (Legacy - File-backed)
  // ============================================================================

  SETTINGS: {
    READ_MAIN_SETTINGS: 'main-settings:read',
    WRITE_MAIN_SETTINGS: 'main-settings:write',
    GET_DEFAULT_DOWNLOAD_PATH: 'main-settings:get-default-download-path',
    GET_SETUP_STATUS: 'get-setup-status',
  },

  // ============================================================================
  // LOGGING
  // ============================================================================

  LOGGING: {
    SET_VERBOSE: 'logging:set-verbose',
    GET_VERBOSE: 'logging:get-verbose',
    LOG: 'logging:log',
  },

  // ============================================================================
  // CACHE MANAGEMENT
  // ============================================================================

  CACHE: {
    CLEAR_SOURCE_CACHE: 'clear-source-cache',
  },

  // ============================================================================
  // UTILITIES
  // ============================================================================

  UTILS: {
    PATH_JOIN: 'path-join',
    GET_ENV_VAR: 'get-env-var',
  },

  // ============================================================================
  // SEARCH (Database and YouTube search)
  // ============================================================================

  SEARCH: {
    DATABASE: 'search:database',
    YOUTUBE: 'search:youtube',
    HISTORY_GET: 'search:history:get',
    CACHED_RESULTS_GET: 'search:cached-results:get',
  },

  // ============================================================================
  // WISHLIST (Video approval workflow)
  // ============================================================================

  WISHLIST: {
    ADD: 'wishlist:add',
    REMOVE: 'wishlist:remove',
    GET_BY_STATUS: 'wishlist:get-by-status',
    APPROVE: 'wishlist:approve',
    DENY: 'wishlist:deny',
    BULK_APPROVE: 'wishlist:bulk-approve',
    BULK_DENY: 'wishlist:bulk-deny',
  },

  // ============================================================================
  // UI / WINDOWS
  // ============================================================================

  UI: {
    OPEN_VIDEO_IN_WINDOW: 'open-video-in-window',
  },

  // ============================================================================
  // TESTING
  // ============================================================================

  TEST: {
    TEST_HANDLER: 'test-handler',
  },
} as const;

/**
 * Type helper to extract all channel strings from the IPC object
 * Useful for validating that all channels are registered
 */
export type IPCChannel = typeof IPC[keyof typeof IPC][keyof typeof IPC[keyof typeof IPC]];

/**
 * Helper function to get all IPC channel strings as a flat array
 * Useful for testing and validation
 */
export function getAllIPCChannels(): string[] {
  const channels: string[] = [];

  for (const category of Object.values(IPC)) {
    for (const channel of Object.values(category)) {
      channels.push(channel);
    }
  }

  return channels;
}

/**
 * Helper function to validate if a string is a valid IPC channel
 */
export function isValidIPCChannel(channel: string): boolean {
  const allChannels = getAllIPCChannels();
  return allChannels.includes(channel);
}
