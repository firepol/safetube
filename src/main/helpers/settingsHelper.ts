import { DatabaseService } from '../services/DatabaseService';
import { getSetting, getSettingsByNamespace } from '../database/queries/settingsQueries';
import { MainSettings } from '../../shared/types';

/**
 * Get YouTube API key from settings with environment variable fallback
 * Priority: Database settings > YOUTUBE_API_KEY environment variable
 */
export async function getYouTubeApiKey(): Promise<string> {
  try {
    const mainSettings = await readMainSettings();
    const apiKey = mainSettings.youtubeApiKey || '';

    if (apiKey) {
      return apiKey;
    }
  } catch (error) {
    console.warn('[SettingsHelper] Could not read API key from database, trying environment variable:', error);
  }

  // Fallback to environment variable
  return process.env.YOUTUBE_API_KEY || '';
}

/**
 * Read main settings from database
 */
export async function readMainSettings(): Promise<MainSettings> {
  try {
    const db = DatabaseService.getInstance();
    const settings = await getSettingsByNamespace(db, 'main');
    return settings as MainSettings;
  } catch (error) {
    console.error('[SettingsHelper] Error reading main settings from database:', error);
    return {};
  }
}

/**
 * Write main settings to database
 */
export async function writeMainSettings(settings: MainSettings): Promise<void> {
  const db = DatabaseService.getInstance();
  const { setSettingsByNamespace } = await import('../database/queries/settingsQueries');
  await setSettingsByNamespace(db, 'main', settings);
}

/**
 * Get a specific setting value with type safety
 */
export async function getSettingValue<T>(key: string, defaultValue?: T): Promise<T | undefined> {
  try {
    const db = DatabaseService.getInstance();
    return await getSetting<T>(db, key, defaultValue);
  } catch (error) {
    console.error(`[SettingsHelper] Error reading setting ${key}:`, error);
    return defaultValue;
  }
}
