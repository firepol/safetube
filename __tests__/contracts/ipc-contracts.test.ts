/**
 * IPC Contract Tests
 *
 * These tests verify that the IPC communication contracts between
 * main, preload, and renderer processes are consistent and complete.
 *
 * What these tests catch:
 * - Missing handler registrations for defined IPC channels
 * - Orphaned handlers (handlers for non-existent channels)
 * - Response format inconsistencies (DatabaseResponse<T> structure)
 * - IPC channel inventory changes (via snapshots)
 */

import { describe, test, expect, beforeAll, vi } from 'vitest';
import { IPC, getAllIPCChannels } from '../../src/shared/ipc-channels';
import { captureIPCHandlers, getRegisteredChannels } from '../helpers/ipc-helpers';

// Mock electron module
vi.mock('electron', () => ({
  ipcMain: {
    handle: vi.fn(),
    on: vi.fn(),
    removeHandler: vi.fn(),
  },
  app: {
    getPath: vi.fn(() => '/mock/path'),
    getName: vi.fn(() => 'SafeTube'),
  },
  BrowserWindow: vi.fn(),
}));

// Mock fs module for handler registration
vi.mock('fs', () => ({
  default: {
    existsSync: vi.fn(() => true),
    readFileSync: vi.fn(() => '{}'),
    writeFileSync: vi.fn(),
    readdirSync: vi.fn(() => []),
    promises: {
      readFile: vi.fn(async () => '{}'),
      writeFile: vi.fn(async () => {}),
    },
  },
}));

// Mock path module
vi.mock('path', () => {
  const mockPath = {
    join: vi.fn((...args) => args.join('/')),
    resolve: vi.fn((...args) => args.join('/')),
    dirname: vi.fn((p) => p.split('/').slice(0, -1).join('/')),
    basename: vi.fn((p) => p.split('/').pop()),
    extname: vi.fn((p) => {
      const parts = p.split('.');
      return parts.length > 1 ? '.' + parts.pop() : '';
    }),
    sep: '/',
    delimiter: ':',
    parse: vi.fn((p) => ({
      root: '',
      dir: p.split('/').slice(0, -1).join('/'),
      base: p.split('/').pop(),
      ext: p.includes('.') ? '.' + p.split('.').pop() : '',
      name: p.split('/').pop()?.split('.')[0] || '',
    })),
    format: vi.fn((pathObject) => pathObject.base || ''),
    isAbsolute: vi.fn((p) => p.startsWith('/')),
    normalize: vi.fn((p) => p),
    relative: vi.fn((from, to) => to),
    toNamespacedPath: vi.fn((p) => p),
  };

  return {
    default: mockPath,
    ...mockPath, // Export both as default and named exports
  };
});

// Mock other Node modules that main process files might import
vi.mock('child_process', () => ({
  spawn: vi.fn(),
  exec: vi.fn(),
}));

vi.mock('crypto', () => ({
  default: {
    randomBytes: vi.fn(() => Buffer.from('mock')),
    createHash: vi.fn(() => ({
      update: vi.fn().mockReturnThis(),
      digest: vi.fn(() => 'mocked-hash'),
    })),
  },
}));

describe('IPC Contract Tests', () => {
  let registeredHandlers: Map<string, Function>;
  let allChannels: string[];

  beforeAll(async () => {
    // Capture handler registrations
    registeredHandlers = captureIPCHandlers();

    // Get all defined IPC channels
    allChannels = getAllIPCChannels();

    // Import and call registerAllHandlers to register all IPC handlers
    // This must happen after mocking electron
    const { registerAllHandlers } = await import('../../src/main/services/ipcHandlerRegistry');
    registerAllHandlers();

    // Also import main.ts to register its handlers (they register at module load)
    // We need to catch errors since these files have side effects
    try {
      await import('../../src/main/main');
    } catch (error) {
      // Ignore errors from importing main process code in test environment
      // We only care about capturing the handler registrations
    }

    // Import index.ts handlers
    try {
      await import('../../src/main/index');
    } catch (error) {
      // Ignore errors, we only need handler registrations
    }
  });

  describe('Channel Registration Validation', () => {
    test('all defined IPC channels have registered handlers', () => {
      const missingHandlers: string[] = [];

      for (const channel of allChannels) {
        if (!registeredHandlers.has(channel)) {
          missingHandlers.push(channel);
        }
      }

      if (missingHandlers.length > 0) {
        console.error('\n❌ Missing handlers for the following channels:');
        missingHandlers.forEach(ch => console.error(`  - ${ch}`));
      }

      expect(missingHandlers).toEqual([]);
    });

    test('no orphaned handlers (handlers for non-existent channels)', () => {
      const validChannels = new Set(allChannels);
      const registeredChannelList = getRegisteredChannels(registeredHandlers);
      const orphanedHandlers: string[] = [];

      for (const channel of registeredChannelList) {
        if (!validChannels.has(channel)) {
          orphanedHandlers.push(channel);
        }
      }

      if (orphanedHandlers.length > 0) {
        console.warn('\n⚠️  Orphaned handlers (not in IPC constants):');
        orphanedHandlers.forEach(ch => console.warn(`  - ${ch}`));
        console.warn('  Consider adding these to src/shared/ipc-channels.ts or removing the handlers');
      }

      expect(orphanedHandlers).toEqual([]);
    });

    test('handler count matches channel count', () => {
      const handlerCount = registeredHandlers.size;
      const channelCount = allChannels.length;

      console.log(`\n📊 IPC Contract Summary:`);
      console.log(`  Total IPC Channels: ${channelCount}`);
      console.log(`  Registered Handlers: ${handlerCount}`);
      console.log(`  Coverage: ${handlerCount === channelCount ? '✅ 100%' : `⚠️  ${Math.round((handlerCount / channelCount) * 100)}%`}`);

      expect(handlerCount).toBe(channelCount);
    });
  });

  describe('Channel Inventory Documentation', () => {
    test('documents IPC channel breakdown by category', () => {
      const breakdown: Record<string, number> = {};

      for (const [category, methods] of Object.entries(IPC)) {
        breakdown[category] = Object.keys(methods).length;
      }

      console.log('\n📋 IPC Channels by Category:');
      for (const [category, count] of Object.entries(breakdown).sort((a, b) => b[1] - a[1])) {
        console.log(`  ${category.padEnd(25)} ${count.toString().padStart(3)} channels`);
      }

      // Snapshot test to detect accidental channel additions/removals
      expect(breakdown).toMatchSnapshot('ipc-channel-breakdown');
    });

    test('all IPC channels snapshot', () => {
      // This will fail if channels are added/removed, alerting developers
      // to update tests and documentation
      expect(allChannels.sort()).toMatchSnapshot('all-ipc-channels');
    });
  });

  describe('Response Format Validation', () => {
    test('database handlers should be prepared to return DatabaseResponse<T> format', () => {
      // Get all database-related channels
      const databaseChannels = [
        ...Object.values(IPC.DATABASE),
        ...Object.values(IPC.FAVORITES),
        ...Object.values(IPC.VIEW_RECORDS),
        ...Object.values(IPC.VIDEOS),
        ...Object.values(IPC.SOURCES),
        ...Object.values(IPC.YOUTUBE_CACHE_DB),
      ];

      // Verify all have handlers registered
      const missingDatabaseHandlers: string[] = [];
      for (const channel of databaseChannels) {
        if (!registeredHandlers.has(channel)) {
          missingDatabaseHandlers.push(channel);
        }
      }

      expect(missingDatabaseHandlers).toEqual([]);
    });
  });

  describe('Naming Convention Validation', () => {
    test('all IPC channels follow naming convention', () => {
      // Expected patterns:
      // - database:category:action
      // - category:action
      // - action (simple handlers)

      const invalidChannels: string[] = [];

      for (const channel of allChannels) {
        // Check for invalid characters
        if (!/^[a-z0-9:-]+$/.test(channel)) {
          invalidChannels.push(`${channel} (contains invalid characters)`);
        }

        // Check for double colons
        if (channel.includes('::')) {
          invalidChannels.push(`${channel} (contains double colons)`);
        }

        // Check for trailing/leading colons
        if (channel.startsWith(':') || channel.endsWith(':')) {
          invalidChannels.push(`${channel} (starts or ends with colon)`);
        }
      }

      if (invalidChannels.length > 0) {
        console.error('\n❌ Invalid channel names:');
        invalidChannels.forEach(ch => console.error(`  - ${ch}`));
      }

      expect(invalidChannels).toEqual([]);
    });

    test('no duplicate channel definitions', () => {
      const channelCounts = new Map<string, number>();

      for (const channel of allChannels) {
        channelCounts.set(channel, (channelCounts.get(channel) || 0) + 1);
      }

      const duplicates = Array.from(channelCounts.entries())
        .filter(([_, count]) => count > 1)
        .map(([channel, count]) => `${channel} (${count} times)`);

      if (duplicates.length > 0) {
        console.error('\n❌ Duplicate channel definitions:');
        duplicates.forEach(dup => console.error(`  - ${dup}`));
      }

      expect(duplicates).toEqual([]);
    });
  });

  describe('Category Organization', () => {
    test('all channels are properly categorized', () => {
      // Ensure no empty categories
      const emptyCategories: string[] = [];

      for (const [category, methods] of Object.entries(IPC)) {
        if (Object.keys(methods).length === 0) {
          emptyCategories.push(category);
        }
      }

      if (emptyCategories.length > 0) {
        console.warn('\n⚠️  Empty IPC categories:');
        emptyCategories.forEach(cat => console.warn(`  - ${cat}`));
      }

      expect(emptyCategories).toEqual([]);
    });

    test('related channels are grouped in same category', () => {
      // This is more of a documentation test
      // Verify that database-related operations are in DATABASE/FAVORITES/etc categories
      const databaseCategories = [
        'DATABASE',
        'FAVORITES',
        'FAVORITES_LEGACY',
        'VIEW_RECORDS',
        'VIDEOS',
        'SOURCES',
        'YOUTUBE_CACHE_DB',
        'YOUTUBE_CACHE',
      ];

      for (const category of databaseCategories) {
        expect(IPC).toHaveProperty(category);
      }
    });
  });
});
