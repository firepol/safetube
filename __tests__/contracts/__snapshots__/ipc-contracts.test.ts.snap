// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IPC Contract Tests > Channel Inventory Documentation > all IPC channels snapshot > all-ipc-channels 1`] = `
[
  "admin:add-extra-time",
  "admin:authenticate",
  "admin:change-password",
  "admin:get-last-watched-video-with-source",
  "admin:get-time-extra",
  "admin:hash-password",
  "admin:write-time-limits",
  "clear-source-cache",
  "database:favorites:add",
  "database:favorites:get-all",
  "database:favorites:is-favorite",
  "database:favorites:remove",
  "database:favorites:toggle",
  "database:health-check",
  "database:settings:get-by-namespace",
  "database:settings:get-setting",
  "database:settings:set-by-namespace",
  "database:settings:set-setting",
  "database:sources:create",
  "database:sources:delete",
  "database:sources:get-all",
  "database:sources:get-by-id",
  "database:sources:update",
  "database:time-limits:get",
  "database:time-limits:get-for-day",
  "database:time-limits:update",
  "database:time-limits:update-day",
  "database:usage-extras:add",
  "database:usage-extras:delete",
  "database:usage-extras:get-by-date",
  "database:usage-extras:get-total-minutes",
  "database:usage-logs:get-by-date",
  "database:usage-logs:get-by-date-range",
  "database:usage-logs:get-monthly",
  "database:usage-logs:increment",
  "database:usage-logs:upsert",
  "database:videos:batch-upsert",
  "database:videos:get-by-id",
  "database:videos:get-by-source",
  "database:videos:search",
  "database:videos:update-availability",
  "database:videos:update-metadata",
  "database:view-records:get",
  "database:view-records:get-history",
  "database:view-records:get-recently-watched",
  "database:view-records:update",
  "database:youtube-cache:clear-cache",
  "database:youtube-cache:get-cached-results",
  "database:youtube-cache:set-cached-results",
  "download:cancel",
  "download:check-downloaded",
  "download:get-status",
  "download:is-downloading",
  "download:reset-status",
  "download:start",
  "downloaded-videos:get-all",
  "downloaded-videos:get-by-id",
  "downloaded-videos:get-by-source",
  "downloaded-videos:get-total-size",
  "downloads:cleanup",
  "get-conversion-status",
  "get-dlna-file",
  "get-env-var",
  "get-existing-converted-video-path",
  "get-folder-video-count",
  "get-local-file",
  "get-local-folder-contents",
  "get-local-source-video-count",
  "get-local-video-duration",
  "get-paginated-videos",
  "get-player-config",
  "get-setup-status",
  "get-video-codec-info",
  "get-video-data",
  "get-video-streams",
  "get-watched-videos",
  "get-youtube-api-key",
  "get-youtube-video-info",
  "has-converted-video",
  "load-all-videos-from-sources",
  "load-sources-for-kid-screen",
  "load-videos-for-source",
  "load-videos-from-sources",
  "logging:get-verbose",
  "logging:log",
  "logging:set-verbose",
  "main-settings:get-default-download-path",
  "main-settings:read",
  "main-settings:write",
  "needs-video-conversion",
  "open-video-in-window",
  "path-join",
  "search:cached-results:get",
  "search:database",
  "search:history:get",
  "search:youtube",
  "start-video-conversion",
  "test-handler",
  "time-tracking:get-time-limits",
  "time-tracking:get-time-tracking-state",
  "time-tracking:record-video-watching",
  "video-sources:get-all",
  "video-sources:save-all",
  "video-sources:validate-local-path",
  "video-sources:validate-youtube-url",
  "wishlist:add",
  "wishlist:approve",
  "wishlist:bulk-approve",
  "wishlist:bulk-deny",
  "wishlist:deny",
  "wishlist:get-by-status",
  "wishlist:remove",
  "youtube-cache:clear-expired",
  "youtube-cache:clear-source",
  "youtube-cache:get",
  "youtube-cache:get-page",
  "youtube-cache:load-config",
  "youtube-cache:save-page",
  "youtube-cache:set",
]
`;

exports[`IPC Contract Tests > Channel Inventory Documentation > documents IPC channel breakdown by category > ipc-channel-breakdown 1`] = `
{
  "ADMIN": 7,
  "CACHE": 1,
  "CONVERSION": 6,
  "DATABASE": 1,
  "DB_SETTINGS": 4,
  "DLNA": 1,
  "DOWNLOADED_VIDEOS": 4,
  "DOWNLOADS": 7,
  "FAVORITES": 5,
  "LOCAL_FILES": 5,
  "LOGGING": 3,
  "PLAYBACK": 2,
  "SEARCH": 4,
  "SETTINGS": 4,
  "SOURCES": 5,
  "TEST": 1,
  "TIME_LIMITS": 4,
  "TIME_TRACKING": 3,
  "UI": 1,
  "USAGE_EXTRAS": 4,
  "USAGE_LOGS": 5,
  "UTILS": 2,
  "VIDEOS": 6,
  "VIDEO_LOADING": 7,
  "VIDEO_SOURCES": 4,
  "VIEW_RECORDS": 4,
  "WISHLIST": 7,
  "YOUTUBE": 2,
  "YOUTUBE_CACHE": 4,
  "YOUTUBE_CACHE_DB": 6,
}
`;
